src/algorithms/time_evolution/simpleupdate.jl

```diff
@@ -98,73 +102,113 @@ function _su_ybond!(
     return ϵ
 end
 
 """
-    su_iter(peps::InfinitePEPS, gate::LocalOperator, alg::SimpleUpdate, env::SUWeight; bipartite::Bool=false)
+    su_iter(peps::P, gate::LocalOperator, alg::SimpleUpdate, env::SUWeight; kwargs...) where {P <: InfiniteState}
 
-One round of simple update on `peps` applying the nearest neighbor `gate`.
-When the input `peps` has a unit cell size of (2, 2), one can set `bipartite = true` to enforce the bipartite structure. 
+One round of simple update on InfinitePEPS or InfinitePEPO `state` applying the nearest neighbor `gate`.
+When the input `state` has a unit cell size of (2, 2), one can set `bipartite = true` to enforce the bipartite structure. 
 """
 function su_iter(
-        peps::InfinitePEPS,
-        gate::LocalOperator,
-        alg::SimpleUpdate,
-        env::SUWeight;
-        bipartite::Bool = false,
+        state::InfinitePEPS, gate::LocalOperator, alg::SimpleUpdate, env::SUWeight;
+        bipartite::Bool = false, gate_side::Symbol = :both
     )
-    @assert size(gate.lattice) == size(peps)
-    Nr, Nc = size(peps)
+    @assert size(gate.lattice) == size(state)
+    @assert gate_side in (:both, :codomain)
+    Nr, Nc = size(state)
     bipartite && (@assert Nr == Nc == 2)
     (Nr >= 2 && Nc >= 2) || throw(
-        ArgumentError("iPEPS unit cell size for simple update should be no smaller than (2, 2)."),
+        ArgumentError("`state` unit cell size for simple update should be no smaller than (2, 2)."),
     )
-    peps2, env2 = deepcopy(peps), deepcopy(env)
+    state2, env2 = deepcopy(state), deepcopy(env)
     for r in 1:Nr, c in 1:Nc
         term = get_gateterm(gate, (CartesianIndex(r, c), CartesianIndex(r, c + 1)))
         trscheme = truncation_scheme(alg.trscheme, 1, r, c)
-        _su_xbond!(peps2, term, env2, r, c, trscheme)
+        _su_xbond!(state2, term, env2, r, c, trscheme)
         if bipartite
             rp1, cp1 = _next(r, Nr), _next(c, Nc)
-            peps2.A[rp1, cp1] = deepcopy(peps2.A[r, c])
-            peps2.A[rp1, c] = deepcopy(peps2.A[r, cp1])
+            state2.A[rp1, cp1] = deepcopy(state2.A[r, c])
+            state2.A[rp1, c] = deepcopy(state2.A[r, cp1])
             env2.data[1, rp1, cp1] = deepcopy(env2.data[1, r, c])
         end
         term = get_gateterm(gate, (CartesianIndex(r, c), CartesianIndex(r - 1, c)))
         trscheme = truncation_scheme(alg.trscheme, 2, r, c)
-        _su_ybond!(peps2, term, env2, r, c, trscheme)
+        _su_ybond!(state2, term, env2, r, c, trscheme)
         if bipartite
             rm1, cm1 = _prev(r, Nr), _prev(c, Nc)
-            peps2.A[rm1, cm1] = deepcopy(peps2.A[r, c])
-            peps2.A[r, cm1] = deepcopy(peps2.A[rm1, c])
+            state2.A[rm1, cm1] = deepcopy(state2.A[r, c])
+            state2.A[r, cm1] = deepcopy(state2.A[rm1, c])
             env2.data[2, rm1, cm1] = deepcopy(env2.data[2, r, c])
         end
     end
-    return peps2, env2
+    return state2, env2
+end
+
+function su_iter(
```

Comment:
Similar comment about the indexing behavior here as well, I think these implementations can be joined that way, to reduce code duplication